cpp/include/raft/core/resource/sub_comms.hpp

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/comms.hpp>
+#include <raft/core/resource/resource_types.hpp>
+#include <raft/core/resources.hpp>
+
+namespace raft::resource {
+class sub_comms_resource : public resource {
+ public:
+  sub_comms_resource() {}
+  void* get_resource() override { return &communicators_; }
+
+  ~sub_comms_resource() override {}
+
+ private:
+  std::unordered_map<std::string, std::shared_ptr<comms::comms_t>> communicators_;
```

Comment:
That's fair. Looking at [this blog](http://scottmeyers.blogspot.com/2015/09/should-you-be-using-something-instead.html) from Scott Meyers, it looks like a linear search was actually the fastest for very small numbers of elements. That being said, an ordered vector and lower_bound doesn't take that much more effort so I think we should just use that.

Reply:
Do we need to use a linear_search? Since the resource_type is a small integer - can't we just allocate a vector with the max number of resources, and then just index into it instead (trading off a small amount of memory space for lookup speed)

Reply:
Do we need to use a linear_search? Since the resource_type is a small integer, we could just allocate a vector with the maximum size of the resource_type - and then index into that instead.

Reply:
> Do we need to use a linear_search? Since the resource_type is a small integer, we could just allocate a vector with the maximum size of the resource_type - and then index into that instead.

I certainly don't mind this approach while we know that the number of potential resources could be quite small. My initial concern is when this object gets used more generically outside of the resource_type enum but we can revisit when we come to that point.